### PR TITLE
feat(cosh): add large paste placeholder and fix placeholder id reset on esc cancel

### PR DESCRIPTION
**Pull request by kongld-mq**
## Description

feat(cosh): add large paste placeholder and fix placeholder id reset on esc cancel

- Create placeholder for pastes >1000 chars or >10 lines instead of full content
- Support sequential IDs for multiple pastes of same size with reuse mechanism
- Atomically delete placeholder on Backspace
- Expand placeholder to full content on submit
- Use pasteWorkaround check for macOS Enter-fix to prevent double-submit
- Handle clipboard errors with debugLogger instead of console.error
- Clear pendingPastes and activePlaceholderIds on double ESC to reset ID counter
- Add 9 new test cases covering placeholder creation, expansion, deletion, reuse, and ESC reset
- Fix enter-submit on macOS: only apply recentPasteTime protection when
  pasteWorkaround is enabled (Windows or Node < 20). On macOS/Linux with
  modern Node, bracketed paste markers work reliably so the protection is
  unnecessary and caused the first Enter after paste to be ignored.

## Related Issue

closes #

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional change)
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Scope

- [ ] `cosh` (copilot-shell)
- [ ] `sec-core` (agent-sec-core)
- [ ] `skill` (os-skills)
- [ ] `sight` (agentsight)
- [ ] Multiple / Project-wide

## Checklist

- [ ] I have read the [Contributing Guide](../CONTRIBUTING.md)
- [ ] My code follows the project's code style
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the documentation accordingly
- [ ] For `cosh`: Lint passes, type check passes, and tests pass
- [ ] For `sec-core` (Rust): `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] For `sec-core` (Python): Ruff format and pytest pass
- [ ] For `skill`: Skill directory structure is valid and shell scripts pass syntax check
- [ ] For `sight`: `cargo clippy -- -D warnings` and `cargo fmt --check` pass
- [ ] Lock files are up to date (`package-lock.json` / `Cargo.lock`)

## Testing

After generating the test text, paste it into the input box
### large text test(>1000 chars)
```
python3 -c "print('A' * 1500)"
```

### Multi-line test (>10 lines)
```
for i in {1..15}; do echo "Line $i"; done
```
### Small text test(<1000 chars)
```
python3 -c "print('A' * 500)"
```
## Additional Notes

大文本粘贴 | 粘贴 1500 字符 | 显示 [Pasted Content 1500 chars]
-- | -- | --
多行粘贴 | 粘贴 15 行 | 显示 [Pasted Content XX chars]
小文本粘贴 | 粘贴 500 字符 | 正常显示完整文本
多次粘贴 | 连续粘贴两次 1001 字符 | 第二次显示 `#2`
Backspace | 粘贴后按 Backspace | 整体删除占位符

### CODE REVIEW

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx [hunk 1]
```diff
@@ -760,6 +864,48 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+          key.sequence === '\x7f' ||
+          (key.ctrl && key.name === 'h'))
+      ) {
+        const text = buffer.text;
+        const [row, col] = buffer.cursor;
+
+        // Calculate the offset where the cursor is
+        let offset = 0;
+        for (let i = 0; i < row; i++) {
+          offset += buffer.lines[i].length + 1; // +1 for newline
```
Comment (samchu-zsl) [inline src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: 🟡 **UTF-16 与 code-point 偏移语义混用**

这里的 `offset` 把两种单位混在了一起：
- `buffer.lines[i].length`（JS `String.length`）返回的是 **UTF-16 code unit** 数
- `buffer.cursor[1]`（即 `col`）在 `text-buffer.ts` 里是通过 `cpLen(...)` 计算出的 **code point** 数（见 `packages/cli/src/ui/components/shared/text-buffer.ts` 的 `offsetToLogicalPos`）
- 下方 L889 `text.slice(placeholderStart, offset)` 走 UTF-16 index，而 L892 `buffer.replaceRangeByOffset(placeholderStart, offset, '')` 内部用 `offsetToLogicalPos` → `cpLen`，期望的是 code-point 偏移

只要用户在占位符之前手输了 emoji 等非 BMP 字符（🚀 等），`cpLen` 和 UTF-16 长度就会差 1，`placeholderStart` 计算出来指向代理对中间，`text.slice(...)` 比较会失败 / `replaceRangeByOffset` 的起止点也会错位。

建议统一语义（两条路任选）：
- 都用 code-point：`offset += cpLen(buffer.lines[i]) + 1;` 并改用 `cpSlice(text, placeholderStart, offset)` 做比较；
- 或都用 UTF-16：接一个 UTF-16 offset → code-point offset 的转换再传给 `replaceRangeByOffset`。

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx [hunk 2]
```diff
@@ -760,6 +864,48 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+              return next;
+            });
+            const parsed = parsePlaceholder(placeholder);
+            if (parsed) {
+              freePlaceholderId(parsed.charCount, parsed.id);
+            }
+            return;
+          }
+        }
+        // No placeholder matched — fall through to buffer's default backspace
```
Comment (samchu-zsl) [inline src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: 🟡 **光标不在占位符末尾时会丢失原文**

这里的 `for` 循环只能匹配「光标恰好位于占位符末尾」一种情况，其他位置都会 fall through 到下面的 `buffer.handleInput(key)` 做默认逐字 backspace。

结果：
1. 用户按 ← 进入占位符字符串内部后按 backspace；
2. 或从占位符后一行回退到占位符所在行末再 backspace；

两种情况下，占位符字符串会被**部分破坏**，但 `pendingPastes` Map 的 key 仍是原完整占位符 → `handleSubmitAndClear` 的正则匹配不到 → 用户最终提交的是残缺的 `[Pasted Content 10...` 字符串，**原始粘贴内容永久丢失且没有任何告警**。

建议对这些分支做明确处理，任一：
- 占位符视为 atomic：只要 backspace 目标交集命中任一 placeholder 的范围，就整体删除；
- 或侦测到占位符被破坏时，同步从 `pendingPastes` 里 `delete(placeholder)` + `freePlaceholderId(...)`，避免状态和 buffer 不一致。

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx [hunk 3]
```diff
@@ -691,7 +793,9 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
       if (keyMatchers[Command.SUBMIT](key)) {
         if (buffer.text.trim()) {
           // Check if a paste operation occurred recently to prevent accidental auto-submission
-          if (recentPasteTime !== null) {
+          // Only apply this protection when pasteWorkaround is enabled (Windows or Node < 20)
+          // On macOS/Linux with modern Node, bracketed paste markers work reliably so the protection is unnecessary
+          if (pasteWorkaround && recentPasteTime !== null) {
```
Comment (samchu-zsl) [inline src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx:865]: 🟡 **gating 变更缺少有效测试覆盖（待作者确认）**

注释说「macOS/Linux 上 modern Node 保护不必要」这个判断本身合理，但这是一个**用户可感知的行为变更**，而现有两条相关测试在 `pasteWorkaround=false`（测试环境默认）下已经不再真正测试保护逻辑：

- `InputPrompt.test.tsx` 的 `should prevent auto-submission immediately after paste with newlines`：因为 `props.buffer.text` 是空的，`buffer.text.trim()` 分支直接短路，根本走不到 L798；测试通过只是因为 onSubmit 从来没被调用。
- `should allow submission after paste protection timeout`：改动后 `pasteWorkaround=false` 让 `wait(600)` 等待本身就无意义。

建议：
1. 新增 `pasteWorkaround=true` 下「粘贴后立即 Enter 被拦截」的回归测试；
2. 或在 `renderWithProviders` 默认开启 `pasteWorkaround=true` 以保留既有语义；
3. CHANGELOG 里把该行为变更对 macOS/Linux 用户显式写一行。

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx [hunk 4]
```diff
@@ -112,13 +119,67 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+    let id = 1;
+    while (activeIds.has(id)) {
+      id++;
+    }
+
+    // Mark as active
+    activeIds.add(id);
+    activePlaceholderIds.current.set(charCount, activeIds);
+
+    const base = `[Pasted Content ${charCount} chars]`;
```
Comment (samchu-zsl) [inline src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: ℹ️ **建议走 i18n**

本文件 L31 已经 `import { t } from '../../i18n/index.js';`，cosh 其它用户可见文案（如 `Type your message or @path/to/file`）都是通过 `t()` + `locales/en.js` / `zh.js` 本地化的，这里直接硬编码英文略显特殊。

建议：
```ts
const base = t('input.paste.placeholder', { charCount: String(charCount) });
```
并在 `locales/en.js` / `locales/zh.js` 补上对应 key（并注意 L141 的正则解析也要能同时匹配中英两种本地化形态，或改用「不进入正则的独立映射」来识别占位符）。

**Review by samchu-zsl — COMMENTED** [review]
## Summary

非常棒的功能扩展 🎉 —— 大粘贴占位符、ID 复用、ESC 重置整体思路清晰，测试用例也覆盖了主要正向路径。以下留下 4 条具体改动建议，按建议处理后整体可以合入：

### 🟡 Major（建议修复）
- **UTF-16 vs code-point 偏移混用**（`InputPrompt.tsx` L878-L892）：backspace 删除占位符的路径里，`buffer.lines[i].length` 与 `text.slice` 走 UTF-16，而 `buffer.cursor` 的 col 和 `replaceRangeByOffset` 走 code-point，混用后只要用户在占位符前输入了 emoji 等非 BMP 字符就会错位。
- **占位符被部分删除后原文静默丢失**（`InputPrompt.tsx` L867-L907）：当前仅匹配「光标恰好在占位符尾」，若用户按 ← 进入占位符内部或从下一行回退再 backspace，占位符字符串被逐字破坏，而 `pendingPastes` 仍持有完整 key，提交时正则匹配失败，用户拿到残缺占位符而非原文。
- **`recentPasteTime` gating 语义变化缺测试**（`InputPrompt.tsx` L798）：加了 `pasteWorkaround &&` 后，原来的 paste-auto-submission 保护测试因为默认 `pasteWorkaround=false`，实际已经不再走到保护分支了；建议补一条 `pasteWorkaround=true` 下的回归测试。

### ℹ️ Minor
- 占位符文案硬编码（`InputPrompt.tsx` L179）：同文件已经 `import { t }`，建议走 i18n。

感谢贡献！

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx (hunk 1 above)
Comment (kongld-mq) [reply src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: Unified offset semantics to use code-point throughout the placeholder backspace deletion logic, and added comprehensive test coverage for emoji/non-BMP scenarios.

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx (hunk 2 above)
Comment (kongld-mq) [reply src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: Use Unicode Private Use Area character \uE000 as single-character marker:

- Placeholder occupies only 1 character position in buffer
- Backspace can only delete this single character (atomic deletion)
- When deleted, corresponding entry is removed from pendingPastes and ID is freed
- Atomic deletion prevents partial corruption entirely

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx (hunk 4 above)
Comment (kongld-mq) [reply src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: Placeholder display text uses t('input.paste.placeholder', { charCount }) for localization
- locales/en.js: '[Pasted Content {{charCount}} chars]'
- locales/zh.js: '[粘贴内容 {{charCount}} 字符]'

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx (hunk 3 above)
Comment (kongld-mq) [reply src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx:865]: Extended renderWithProviders to support pasteWorkaround parameter
Rewrote test cases using large paste (>1000 chars) to trigger placeholder insertion, avoiding issue where handleInput mock doesn't update buffer.text
Mocked buffer.insert to correctly simulate real behavior (appending placeholder to existing text)
Tests added/modified:
﻿
should prevent auto-submission immediately after paste when pasteWorkaround is enabled - verifies gating blocks submission
should allow submission immediately after paste when pasteWorkaround is disabled - verifies normal submission when gating inactive
should not interfere with normal Enter key submission when no recent paste - preserved original test

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
## Review Summary

整体设计思路清晰 —— 用单个 PUA 字符（`\uE000`）作为 marker 存储于 buffer、显示时展开为本地化文案，避免了多字符占位符带来的光标偏移问题。samchu-zsl 提出的 backspace 路径 UTF-16/code-point 混用问题在本次 force push 中已通过引入 \`buffer.offset\` + \`toCodePoints\` 修复，值得肯定。

以下是我的审查意见：

---

### 🔴 Blocker（必须修复）

**1. 渲染循环中第二段计数仍混用 UTF-16 索引**
\`InputPrompt.tsx\` \`placeholderIndexInLine\` 第二段循环：
\`\`\`ts
for (let i = 0; i < logicalStartCol && i < logicalLine.length; i++) {
  if (logicalLine[i] === PLACEHOLDER_MARKER) {
\`\`\`
\`logicalLine[i]\` 使用的是 JS 字符串的 UTF-16 索引，而 \`logicalStartCol\` 来自 \`visualToLogicalMap\`，是 code-point 偏移。当这一行在 \`logicalStartCol\` 之前存在 emoji（如 \`🚀\uE000\`）时，\`i\` 每步增量为 1 个 UTF-16 单元（emoji 占 2 个），导致循环提前终止，placeholder 计数少算，后续 \`pendingPastes[placeholderIndexInLine]\` 取到的是错误条目，显示的占位符文案与实际粘贴内容不匹配。

**修复建议**：将第二段改为 code-point 迭代：
\`\`\`ts
const cpLine = toCodePoints(logicalLine);
for (let i = 0; i < logicalStartCol && i < cpLine.length; i++) {
  if (cpLine[i] === PLACEHOLDER_MARKER) {
    placeholderIndexInLine++;
  }
}
\`\`\`

---

**2. \`Ctrl+C\`（CLEAR_INPUT）未清理 \`pendingPastes\`**
\`InputPrompt.tsx\`：
\`\`\`ts
if (keyMatchers[Command.CLEAR_INPUT](key)) {
  if (buffer.text.length > 0) {
    buffer.setText('');
    resetCompletionState();  // ← 没有 setPendingPastes([])
  }
  return;
}
\`\`\`
Ctrl+C 清空 buffer 但不清空 \`pendingPastes\`，之后再次粘贴大文本时，新 marker 会被 append 到已有的 stale entries 后面（\`setPendingPastes(prev => [...prev, ...])\`），导致新 marker 在渲染时匹配到旧的 paste 条目，显示错误的字符数和内容。

提交时情况更糟：如果 Ctrl+C 后没有重新粘贴，pendingPastes 仍有 stale 条目，\`handleSubmitAndClear\` 中 \`placeholderIdx\` 计数会与 buffer 中实际 marker 个数不一致，带来未定义行为。

**修复建议**：在 CLEAR_INPUT 处理中同步清理：
\`\`\`ts
buffer.setText('');
setPendingPastes([]);
activePlaceholderIds.current.clear();
resetCompletionState();
\`\`\`

---

**3. \`killLineRight\`、\`killLineLeft\`、\`deleteWordLeft/Right\` 删除 marker 后 \`pendingPastes\` 未同步**

这些操作可以删除 buffer 中的 PLACEHOLDER_MARKER，但均未触发 pendingPastes 更新（orphan sync 只在 backspace 路径执行）。用户按 Ctrl+K kill 掉包含 marker 的行尾后，pendingPastes 仍持有该 entry。提交时 buffer 中没有对应 marker，该次粘贴的内容被静默丢弃，用户看不到任何提示。

**修复建议**：在这些操作执行前/后做一次 marker 同步，或把 orphan 清理逻辑抽成公共函数，在每次 handleInput 返回前统一调用。

---

### 🟡 Major（建议修复）

**4. 缺少 \`pasteWorkaround=true\` 下的回归测试**（samchu-zsl 已提，force push 后仍未补充）

\`pasteWorkaround &&\` 修改使原有的 paste-auto-submission 保护在默认测试环境（\`pasteWorkaround=false\`）下完全不生效。需要补充一条 \`pasteWorkaround=true\` 场景下的测试：粘贴后紧接着触发 Enter，验证不会自动提交。

---

**5. 空测试用例**
\`text-buffer.test.ts\`：
\`\`\`ts
it('should correctly slice text with emoji using code-point offsets', () => {
  // This test is documented in textUtils.test.ts (cpSlice tests)
  // ...
  // 无任何 expect 断言
});
\`\`\`
没有断言的测试永远通过，形同虚设。建议补充实际断言或删除。

---

### ℹ️ Minor

**6. \`#${id}\` 后缀未走 i18n**

\`placeholderToLocalized\` 中：
\`\`\`ts
return id === 1 ? baseLocalized : \`${baseLocalized} #${id}\`;
\`\`\`
基础文案已走 \`t()\`，但 \`#2\`、\`#3\` 后缀硬编码为英文。建议在 locale 文件中增加 \`input.paste.placeholder.numbered\` key，或至少保持和基础文案在同一个模板变量中处理。

**7. \`PLACEHOLDER_MARKER = '\uE000'\` 与用户粘贴内容的潜在冲突**

\`\uE000\` 是 Unicode PUA 区的第一个字符，理论上用户可以在外部编辑器中输入或复制包含此字符的内容，触发意外的 placeholder 渲染和删除逻辑。建议在代码注释中说明这一假设，并考虑是否在外部编辑器回写时做过滤。

**Review by kongche-jbw — DISMISSED** [review]
## PR #312 Review: feat(cosh): add large paste placeholder and fix placeholder id reset on esc cancel

### Positive Aspects

- **Good architectural choice**: Using a single PUA character (``) as the marker is elegant — cursor movement, selection, and undo all work naturally since the marker is 1 code point.
- **Thorough test coverage**: 9 new test cases covering placeholder creation, expansion, deletion, ID reuse, ESC reset, emoji interactions, and edge cases.
- **Proper Unicode handling**: Consistent use of code-point semantics (`cpLen`, `cpSlice`, `toCodePoints`) throughout, which avoids surrogate pair bugs.
- **Good i18n support**: Placeholder text is localized via `en.js`/`zh.js` with template interpolation.
- **macOS fix is well-scoped**: Gating `recentPasteTime` behind `pasteWorkaround` is the correct fix — bracketed paste markers work reliably on macOS/Linux with modern Node.

---

### Issues & Suggestions

#### 1. [Critical] `syncPendingPastesWithBuffer` assumes deletion order is always tail-first

```typescript
// InputPrompt.tsx — syncPendingPastesWithBuffer
if (currentPendingPastes.length > markerCount) {
  const excessEntries = currentPendingPastes.slice(markerCount);  // ← always trims from end
  ...
  const syncedPendingPastes = currentPendingPastes.slice(0, markerCount);
```

This assumes excess entries are always at the **end** of the array. But if a user deletes the **first** placeholder marker (via cursor movement + backspace), the first entry should be removed, not the last. The `syncPendingPastesWithBuffer` would then incorrectly keep entry #1 (deleted) and discard entry #2 (still present), causing submit to expand the wrong content.

The explicit backspace handler does handle this correctly by using `deleteIndex`, but `killLineRight`, `killLineLeft`, and `deleteWordLeft` all call `syncPendingPastesWithBuffer` which has this bug.

**Suggestion**: Either scan the buffer to find which markers remain and match them positionally, or track markers by position rather than array index.

#### 2. [Medium] `pendingPastes` in `handleSubmitAndClear` dependencies causes unnecessary recreation

```typescript
// InputPrompt.tsx — handleSubmitAndClear deps
[..., pendingPastes],
```

Every paste updates `pendingPastes`, which recreates `handleSubmitAndClear`. This callback is passed to `useKeypress`, which will unsubscribe/re-subscribe the keypress handler on every paste. Consider using `pendingPastesRef.current` inside the callback and removing `pendingPastes` from the dependency array (similar to the pattern already used in `syncPendingPastesWithBuffer`).

#### 3. [Medium] Skipped test hides regression risk

```typescript
it.skip('should allow submission after paste protection timeout when pasteWorkaround is enabled', ...
```

The comment says "testing React state updates via setTimeout is problematic in the test environment." This is a critical path — if the timeout-based clearing of `recentPasteTime` regresses, users on Windows/old Node will be permanently blocked from submitting after paste. Consider using `vi.useFakeTimers()` + `vi.advanceTimersByTime(600)` instead of real `setTimeout`.

#### 4. [Nit] Duplicate comment block

```typescript
// Handle backspace with placeholder-aware deletion
// Since placeholder marker is a single character (PLACEHOLDER_MARKER),
// we just need to check if backspace would delete a marker.
// If so, also remove the corresponding entry from pendingPastes.
// Handle backspace with placeholder-aware deletion     ← duplicate
```

Lines around the backspace handler have a duplicate block comment. Remove one.

#### 5. [Low] `PLACEHOLDER_MARKER` collision — no runtime guard for external editor

The doc comment correctly identifies the risk of `` appearing in user content via external editor. Consider adding a guard in the external editor callback path to strip or escape ``.

#### 6. [Low] `toCodePoints` called repeatedly in hot render path

In the render loop, `toCodePoints(logicalLine)` is called for each visual line on every render to count placeholder markers. For large inputs with many lines, consider a simpler string iteration to avoid array allocation.

#### 7. [Nit] package-lock.json: Unrelated `peer: true` removals

~20 `"peer": true` removals inflate the diff by ~166 lines. Consider splitting into a separate commit or documenting why.

#### 8. [Low] Box layout change (`height={1}` → `minHeight={1}` + `wrap="wrap"`) may affect non-placeholder line rendering

Verify long non-placeholder lines still render correctly — a line that wraps could now take 2+ rows instead of being truncated.

#### 9. [Nit] PR template not filled in

`closes #` is empty, checklist items unchecked. Please link the related issue and check the appropriate boxes.

---

### Summary

The core feature design is solid — using a PUA character as an atomic marker is a clean approach. The main concern is the **positional sync bug in `syncPendingPastesWithBuffer`** (issue #1), which could cause wrong content expansion when users delete non-trailing placeholders via kill-line or word-delete commands. Issues #2 and #3 are secondary but worth addressing before merge.

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx [hunk 5]
```diff
@@ -132,6 +141,99 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+      }
+    }
+
+    // If pendingPastes has more entries than markers, trim excess (orphan entries)
+    if (currentPendingPastes.length > markerCount) {
+      const excessEntries = currentPendingPastes.slice(markerCount);
+      for (const entry of excessEntries) {
+        freePlaceholderId(entry.charCount, entry.id);
+      }
+      const syncedPendingPastes = currentPendingPastes.slice(0, markerCount);
```
Comment (samchu-zsl) [inline src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: 🔴 Bug: `syncPendingPastesWithBuffer` incorrectly maps entries when a non-tail marker is deleted

When `killLineLeft`/`killLineRight`/`deleteWordLeft` removes a marker that isn't the last one in the buffer, the sync logic trims from the **end** of `pendingPastes`:

```ts
const syncedPendingPastes = currentPendingPastes.slice(0, markerCount);
```

But the surviving markers may correspond to entries at different positions in the original array.

**Repro scenario:**
1. Paste large text A → buffer: `MARKER_A`, pendingPastes: `[{contentA, id:1}]`
2. Type "middle"
3. Paste large text B → buffer: `MARKER_A middle MARKER_B`, pendingPastes: `[{contentA}, {contentB}]`
4. Move cursor to "middle", press `Ctrl+U` (killLineLeft) → buffer becomes `"middle MARKER_B"` (1 marker remaining)
5. Sync sees 1 marker, keeps `pendingPastes.slice(0, 1)` = `[{contentA}]`
6. On submit, the remaining `MARKER_B` expands to `contentA` instead of `contentB`

**Suggested fix:** Track which markers survived (by position in the new buffer text) and keep only the corresponding `pendingPastes` entries, rather than blindly slicing from the start.

**Review by samchu-zsl — CHANGES_REQUESTED** [review]
## Summary

Large paste placeholder feature looks solid for the primary use case. One correctness issue found in the sync logic.

### 🔴 Blocker

- `InputPrompt.tsx:~209` — `syncPendingPastesWithBuffer` uses positional `slice(0, markerCount)` which maps wrong entries when a non-tail marker is deleted via kill-line commands. Results in wrong content expansion on submit.

File: src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx (hunk 5 above)
Comment (kongld-mq) [reply src/copilot-shell/packages/cli/src/ui/components/InputPrompt.tsx]: Fixed by passing the explicit deleted range to syncPendingPastesWithBuffer instead of blindly slicing from the tail.
Before:
```
syncPendingPastesWithBuffer(); // count markers, slice(0, markerCount)
```
after:
```
const oldText = buffer.text;
const cursorOffset = buffer.offset;
// compute deleted range [delStartCp, delEndCp) based on operation type
buffer.killLineLeft();
syncPendingPastesWithBuffer(oldText, delStartCp, delEndCp);
```
The function now counts markers before and inside the deleted range in the old text, then splices exactly those entries from pendingPastes. This correctly handles non-tail marker deletion.

killLineLeft → range = [lineStart, cursor)
killLineRight → range = [cursor, lineEnd)
deleteWordLeft → range = [newCursor, oldCursor)
The count-based fallback (no args) is preserved as a safety-net for the backspace path, which already handles precise marker deletion via deleteIndex.

Added regression tests for both killLineLeft and killLineRight scenarios. deleteWordLeft uses the same sync logic so coverage is equivalent.

**Review by samchu-zsl — APPROVED** [review]
LGTM, thx!
